pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,795 @@
+	grafana, err := h.deploymentCache.Get(CattleMonitoringSystemNamespace, RancherMonitoringGrafana)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("grafana no found. skip")
+			return nil
+		}
+		return fmt.Errorf("grafana get error %v", err)
+	}
+
+	logrus.Infof("current Grafana replicas: %v", *grafana.Spec.Replicas)
+	// check stopped or not
+	if *grafana.Spec.Replicas != 0 {
+		logrus.Infof("stop grafana")
+		grafanaCopy := grafana.DeepCopy()
+		grafanaCopy.Annotations[ReplicaStorageNetworkAnnotation] = strconv.Itoa(int(*grafana.Spec.Replicas))
+		*grafanaCopy.Spec.Replicas = 0
+
+		if _, err := h.deployments.Update(grafanaCopy); err != nil {
+			return fmt.Errorf("Grafana update error %v", err)
+		}
+		return nil
+	}
+
+	return nil
+}
+
+func (h *Handler) checkVMImportControllerStatusAndStop() error {
+	// check deployment harvester-system/harvester-harvester-vm-import-controller replica
+	vmimportcontroller, err := h.deploymentCache.Get(HarvesterSystemNamespace, HarvesterVMImportController)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("VM import controller no found. skip")
+			return nil
+		}
+		return fmt.Errorf("vmimportcontroller get error %v", err)
+	}
+
+	// check stopped or not
+	if *vmimportcontroller.Spec.Replicas != 0 {
+		logrus.Infof("current VM Import Controller replicas: %v", *vmimportcontroller.Spec.Replicas)
+		logrus.Infof("stop vmi import controller")
+		vmimportcontrollerCopy := vmimportcontroller.DeepCopy()
+		vmimportcontrollerCopy.Annotations[ReplicaStorageNetworkAnnotation] = strconv.Itoa(int(*vmimportcontroller.Spec.Replicas))
+		*vmimportcontrollerCopy.Spec.Replicas = 0
+
+		if _, err := h.deployments.Update(vmimportcontrollerCopy); err != nil {
+			return fmt.Errorf("VM Import Controller update error %v", err)
+		}
+		return nil
+	}
+
+	return nil
+}
+
+// check Pod status, if all pods are stopped, return true
+func (h *Handler) checkPodStatusAndStop() error {
+	if err := h.checkRancherMonitoringStatusAndStop(); err != nil {
+		return err
+	}
+
+	if err := h.checkPrometheusStatusAndStop(); err != nil {
+		return err
+	}
+
+	if err := h.checkAltermanagerStatusAndStop(); err != nil {
+		return err
+	}
+
+	if err := h.checkGrafanaStatusAndStop(); err != nil {
+		return err
+	}
+
+	if err := h.checkVMImportControllerStatusAndStop(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (h *Handler) getLonghornStorageNetwork() (string, error) {
+	storage, err := h.longhornSettingCache.Get(util.LonghornSystemNamespaceName, longhornStorageNetworkName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return "", nil
+		}
+		return "", err
+	}
+	return storage.Value, nil
+}
+
+func (h *Handler) updateLonghornStorageNetwork(storageNetwork string) error {
+	storage, err := h.longhornSettingCache.Get(util.LonghornSystemNamespaceName, longhornStorageNetworkName)
+	if err != nil {
+		return err
+	}
```

Comment:
Do we want to handle `IsNotFound` error here? A nonexistent longhorn storage-network setting can pass `getLonghornStorageNetwork` function.

Reply:
Guangbo said that should be always there
If LH setting is not existing, it needs manual config.

Reply:
Got it. Thank you. 👍